Drop inherited attrs and set metadata attrs on estimator results

aazuspan opened the pull request:
This refines how attrs are tracked in Xarray objects returned by spatial estimators. Specifically:

1. The attrs of the input data are no longer inherited by the output data (except for the coordinate attrs which contain the spatial reference). This avoids retaining attrs that are no longer accurate, e.g. storing the `scale_factor` or `unit` of a feature on a predicted target.
2. New metadata attrs `long_name` and `source_method` are set on the output. `long_name` is part of the CF convention for describing variable names. `source_method` is *not* a standard attr, but I thought it might be useful to track provenance by recording the estimator class and method used to produce the output, e.g. `RandomForestRegressor.predict`. These are in addition to `_FillValue` which we already set.

I also had to make an unrelated change to the CI workflow that lists the test dependency versions in order to track down a bug in an old xarray version.

@grovduck, attrs are something I hadn't put much thought into before, so this is an attempt to clean things up and roughly follow CF conventions, since I think that's what would generally be expected by Xarray users. I've been using [this](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.12/cf-conventions.html#attribute-appendix) and [this](https://docs.unidata.ucar.edu/netcdf-c/current/attribute_conventions.html) as references for attributes. 

There are a couple decisions I wasn't sure on that I could use your input:

1. Should we add a `keep_attrs=False` parameter to every estimator method to allow inheriting the input attrs in the output? That's slightly complicated since I think we would still want to override `_FillValue` and `long_name`, but otherwise I think the only downside is having to document another parameter.
2. I'm storing the wrapped method name in a custom attr `source_method`, but we *could* store it in `comment` or `history`, which are standard attrs that seem to generally fit this usage. I think the downside to using those is 1) users are more likely to want to edit them and 2) if we add a `keep_attrs` parameter we probably need to append to them rather than replacing them as they may already exist. The upside of course is that they're already well-defined and should be recognizable.

Any feedback appreciated!

grovduck commented on src/sklearn_raster/estimator.py:
```diff
@@ -618,9 +627,16 @@ def inverse_transform(
             allow_cast=allow_cast,
             check_output_for_nodata=check_output_for_nodata,
             nan_fill=0.0,
+            set_attrs={
+                "source_method": self._method_name(wrapped_func),
+            },
```
Super minor, but if you remove the comma at the end, I think this fits on one line (like all the rest)

grovduck commented on src/sklearn_raster/estimator.py:
```diff
@@ -216,6 +217,7 @@ def predict(
             allow_cast=allow_cast,
             check_output_for_nodata=check_output_for_nodata,
             nan_fill=0.0,
+            set_attrs={"source_method": self._method_name(wrapped_func)},
```
I'm guessing you probably want to remain explicit here, which seems totally reasonable, but a helper function to set "set_attrs" could be used to reduce duplication?
```
def _set_attrs(self, func):
    return {"source_method": self._method_name(func)}
```
and
```
set_attrs=self._set_attrs(wrapped_func)
```
Probably only useful if you foresee adding any more attrs?

aazuspan replied:
I liked this suggestion, but I ended up removing this code when I switched to using the `history` attribute.

grovduck commented on src/sklearn_raster/features.py:
```diff
@@ -214,19 +222,23 @@ def _postprocess_ufunc_output(
         # Transpose features from the last to the first dimension
         result = result.transpose(result.dims[-1], ...)
-
+        # Drop top-level attributes from the input data, but retain coordinate attrs to
+        # preserve the spatial reference, if present. Note that we avoid using
+        # drop_attrs(deep=False) for backwards compatibility due to
+        # https://github.com/pydata/xarray/issues/10027
+        result.attrs = {}
+        if set_attrs is not None:
+            result.attrs.update(set_attrs)
```
This demonstrates my lack of familiarity with `xarray`, but it sound like the coordinate attrs are associated with coordinate objects whereas "regular" attrs would be the ones that get dropped here, is that right?

grovduck reviewed (approved):
@aazuspan, probably not as thorough of a review as you'd like as I'm pretty unfamiliar with metadata in xarray, but it looks like a really cool feature (that I should know much better!).  But your changes look solid and I only had a couple of minor comments.

> Should we add a keep_attrs=False parameter to every estimator method to allow inheriting the input attrs in the output? That's slightly complicated since I think we would still want to override _FillValue and long_name, but otherwise I think the only downside is having to document another parameter.

I'm trying to imagine the scenario where a user would want to keep the attrs from the input features?  The only thing that I can see would be the `history` attr where it sounds like best-practice is to append the current operation to any existing history?  From the page you linked:

> history: A global attribute for an audit trail. This is a character array with a line for each invocation of a program that has modified the dataset. Well-behaved generic netCDF applications should append a line containing: date, time of day, user name, program name and command arguments.

Are there other attrs from the input features that would be useful to retain?

> I'm storing the wrapped method name in a custom attr `source_method`, but we could store it in `comment` or `history`, which are standard attrs that seem to generally fit this usage. I think the downside to using those is 1) users are more likely to want to edit them and 2) if we add a `keep_attrs` parameter we probably need to append to them rather than replacing them as they may already exist. The upside of course is that they're already well-defined and should be recognizable.

A great question.  If the intent here is to conform to the CF convention, is there any risk to introducing custom attrs (e.g. will subsequent software ignore such attrs?).  I'm intrigued by the `history` attr as a place to store this information, but I'm curious about your comment about users wanting to edit them.  By using `history` to store what's curently in `source_method`, are you saying that users may overwrite changes here?  Or that they might want to programmatically use this attr for another purpose?

Again, you are definitely the one with way more knowledge on this, so I defer to what you think is best.

aazuspan commented:
Thanks for the feedback, @grovduck!

>  I'm intrigued by the history attr as a place to store this information, but I'm curious about your comment about users wanting to edit them. By using history to store what's curently in source_method, are you saying that users may overwrite changes here? Or that they might want to programmatically use this attr for another purpose?

I think I was considering that users might overwrite those attrs, but the page you quoted is pretty clear that `history` should be set programatically. I'm convinced that that's a better place to store the method name than `source_method`, so I did some refactoring to 1) preserve any existing history and 2) append our method to the end of history instead of using `source_method`. 

> Are there other attrs from the input features that would be useful to retain?

Good question. Just for reference, here are the attrs from an example NetCDF I pulled [from here](https://www.unidata.ucar.edu/software/netcdf/examples/files.html):

```
{'CVS_Id': '$Id$',
 'creation_date': '',
 'prg_ID': 'Source file unknown Version unknown Date unknown',
 'cmd_ln': 'bds -x 256 -y 128 -m 23 -o [/data/zender/data/dst_T85.nc](https://file+.vscode-resource.vscode-cdn.net/data/zender/data/dst_T85.nc)',
 'history': 'Tue Oct 25 15:08:51 2005: ncks -O -x -v va -m sresa1b_ncar_ccsm3_0_run1_200001.nc sresa1b_ncar_ccsm3_0_run1_200001.nc\nTue Oct 25 15:07:21 2005: ncks -d time,0 sresa1b_ncar_ccsm3_0_run1_200001_201912.nc sresa1b_ncar_ccsm3_0_run1_200001.nc\nTue Oct 25 13:29:43 2005: ncks -d time,0,239 sresa1b_ncar_ccsm3_0_run1_200001_209912.nc [/var/www/html/tmp/sresa1b_ncar_ccsm3_0_run1_200001_201912.nc](https://file+.vscode-resource.vscode-cdn.net/var/www/html/tmp/sresa1b_ncar_ccsm3_0_run1_200001_201912.nc)\nThu Oct 20 10:47:50 2005: ncks -A -v va [/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/sresa1b_ncar_ccsm3_0_run1_va_200001_209912.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/sresa1b_ncar_ccsm3_0_run1_va_200001_209912.nc) [/data/brownmc/sresa1b/atm/mo/tas/ncar_ccsm3_0/run1/sresa1b_ncar_ccsm3_0_run1_200001_209912.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/tas/ncar_ccsm3_0/run1/sresa1b_ncar_ccsm3_0_run1_200001_209912.nc)\nWed Oct 19 14:55:04 2005: ncks -F -d time,01,1200 [/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/sresa1b_ncar_ccsm3_0_run1_va_200001_209912.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/sresa1b_ncar_ccsm3_0_run1_va_200001_209912.nc) [/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/sresa1b_ncar_ccsm3_0_run1_va_200001_209912.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/sresa1b_ncar_ccsm3_0_run1_va_200001_209912.nc)\nWed Oct 19 14:53:28 2005: ncrcat [/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/foo_05_1200.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/foo_05_1200.nc) [/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/foo_1192_1196.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/foo_1192_1196.nc) [/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/sresa1b_ncar_ccsm3_0_run1_va_200001_209912.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/sresa1b_ncar_ccsm3_0_run1_va_200001_209912.nc)\nWed Oct 19 14:50:38 2005: ncks -F -d time,05,1200 [/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/va_A1.SRESA1B_1.CCSM.atmm.2000-01_cat_2099-12.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/va_A1.SRESA1B_1.CCSM.atmm.2000-01_cat_2099-12.nc) [/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/foo_05_1200.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/foo_05_1200.nc)\nWed Oct 19 14:49:45 2005: ncrcat [/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/va_A1.SRESA1B_1.CCSM.atmm.2000-01_cat_2079-12.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/va_A1.SRESA1B_1.CCSM.atmm.2000-01_cat_2079-12.nc) [/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/va_A1.SRESA1B_1.CCSM.atmm.2080-01_cat_2099-12.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/va_A1.SRESA1B_1.CCSM.atmm.2080-01_cat_2099-12.nc) [/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/va_A1.SRESA1B_1.CCSM.atmm.2000-01_cat_2099-12.nc](https://file+.vscode-resource.vscode-cdn.net/data/brownmc/sresa1b/atm/mo/va/ncar_ccsm3_0/run1/va_A1.SRESA1B_1.CCSM.atmm.2000-01_cat_2099-12.nc)\nCreated from CCSM3 case b30.040a\n by wgstrand@ucar.edu\n on Wed Nov 17 14:12:57 EST 2004\n \n For all data, added IPCC requested metadata',
 'table_id': 'Table A1',
 'title': 'model output prepared for IPCC AR4',
 'institution': 'NCAR (National Center for Atmospheric \nResearch, Boulder, CO, USA)',
 'source': 'CCSM3.0, version beta19 (2004): \natmosphere: CAM3.0, T85L26;\nocean     : POP1.4.3 (modified), gx1v3\nsea ice   : CSIM5.0, T85;\nland      : CLM3.0, gx1v3',
 'contact': 'ccsm@ucar.edu',
 'project_id': 'IPCC Fourth Assessment',
 'Conventions': 'CF-1.0',
 'references': 'Collins, W.D., et al., 2005:\n The Community Climate System Model, Version 3\n Journal of Climate\n \n Main website: http://www.ccsm.ucar.edu/',
 'acknowledgment': " Any use of CCSM data should acknowledge the contribution\n of the CCSM project and CCSM sponsor agencies with the \n following citation:\n 'This research uses data provided by the Community Climate\n System Model project ([www.ccsm.ucar.edu](https://file+.vscode-resource.vscode-cdn.net/home/az/code/sklearn-raster/www.ccsm.ucar.edu)), supported by the\n Directorate for Geosciences of the National Science Foundation\n and the Office of Biological and Environmental Research of\n the U.S. Department of Energy.'\nIn addition, the words 'Community Climate System Model' and\n 'CCSM' should be included as metadata for webpages referencing\n work using CCSM data or as keywords provided to journal or book\npublishers of your manuscripts.\nUsers of CCSM data accept the responsibility of emailing\n citations of publications of research using CCSM data to\n ccsm@ucar.edu.\nAny redistribution of CCSM data must include this data\n acknowledgement statement.",
 'realization': np.int32(1),
 'experiment_id': '720 ppm stabilization experiment (SRESA1B)',
 'comment': 'This simulation was initiated from year 2000 of \n CCSM3 model run b30.030a and executed on \n hardware cheetah.ccs.ornl.gov. The input external forcings are\nozone forcing    : A1B.ozone.128x64_L18_1991-2100_c040528.nc\naerosol optics   : AerosolOptics_c040105.nc\naerosol MMR      : AerosolMass_V_128x256_clim_c031022.nc\ncarbon scaling   : carbonscaling_A1B_1990-2100_c040609.nc\nsolar forcing    : Fixed at 1366.5 W m-2\nGHGs             : ghg_ipcc_A1B_1870-2100_c040521.nc\nGHG loss rates   : noaamisc.r8.nc\nvolcanic forcing : none\nDMS emissions    : DMS_emissions_128x256_clim_c040122.nc\noxidants         : oxid_128x256_L26_clim_c040112.nc\nSOx emissions    : SOx_emissions_A1B_128x256_L2_1990-2100_c040608.nc\n Physical constants used for derived data:\n Lv (latent heat of evaporation): 2.501e6 J kg-1\n Lf (latent heat of fusion     ): 3.337e5 J kg-1\n r[h2o] (density of water      ): 1000 kg m-3\n g2kg   (grams to kilograms    ): 1000 g kg-1\n \n Integrations were performed by NCAR and CRIEPI with support\n and facilities provided by NSF, DOE, MEXT and ESC/JAMSTEC.',
 'model_name_english': 'NCAR CCSM'}
```

It looks to me like there are some attributes that you might want to keep, so maybe we should add a parameter to that effect? With the refactor, it should be as simple as just passing that parameter through to the postprocessing method. By defaulting to `False`, I think we put the responsibility on the user to ensure attributes are still accurate.

grovduck commented:
> It looks to me like there are some attributes that you might want to keep, so maybe we should add a parameter to that effect? With the refactor, it should be as simple as just passing that parameter through to the postprocessing method. By defaulting to `False`, I think we put the responsibility on the user to ensure attributes are still accurate.

That seems like a reasonable approach.  At this point, it looks like you've impletmented this for the `_get_attrs` method, but you still need a way for the user to set `preserve_attrs=True` for post-processing?  I think that's what you're saying.

The only other thing I noticed is that `_FillValue` is now an attr on both the Dataset and component DataArrays.  Is that common practice?  Can component DataArrays have different fill values within a Dataset?

aazuspan commented:
> At this point, it looks like you've impletmented this for the _get_attrs method, but you still need a way for the user to set preserve_attrs=True for post-processing? I think that's what you're saying.

Exactly, should be pretty straightforward.

> The only other thing I noticed is that _FillValue is now an attr on both the Dataset and component DataArrays. Is that common practice?

Great point. I think `_FillValue` should exclusively be a variable-level attr, so it shouldn't be set on the Dataset. I'll drop that.

aazuspan commented on .github/workflows/ci.yaml:
```diff
@@ -38,6 +38,11 @@ jobs:
 
+      - name: Install test environment
+        run: |
+          hatch env create test
```
`hatch -e run pip freeze` syncs the env

grovduck commented on src/sklearn_raster/features.py:
```diff
@@ -294,7 +294,7 @@ def _get_attrs(
 
         if keep_attrs:
-            return set_attrs | attrs
+            return attrs | set_attrs
```
Sorry, I totally missed this one when reviewing as well.  This makes sense, though.

aazuspan replied:
No worries! I just learned about this feature, but apparently didn't learn to use it correctly 😅

grovduck reviewed (approved):
@aazuspan, LGTM!  One comment in there about wording, but feel free to ignore.

aazuspan commented on src/sklearn_raster/features.py:
```diff
@@ -107,6 +108,8 @@ def ufunc(x):
             output_core_dims=output_dims,
             output_dtypes=output_dtypes,
+            # Always keep_attrs here to avoid dropping coordinate atttrs. Unwanted attrs
+            # will be dropped during postprocessing.
```
That was my intent, but I can see how the wording was confusing. I expanded the explanation a little bit to:

```
# Keep all attributes here to avoid dropping the spatial reference from the 
# coordinate attributes. Unwanted attrs will be dropped during 
# postprocessing.
```

aazuspan commented:
Thanks for the reviews, @grovduck! Looks like everything is resolved, so I'll go ahead and merge.
